### PR TITLE
feat: Set `Stream._MAX_RECORDS_LIMIT ` during tap testing

### PR DESCRIPTION
**Pull request by kgpayne**
This currently only works with `SQLStream` instances, which apply the `max_records_limit` in `get_records()` by default. For other types of `Stream`, setting a `max_records_limit` in `SuiteConfig` will cause the tests to fail. See #1349 for more details.

----
:books: Documentation preview :books:: https://meltano-sdk--1399.org.readthedocs.build/en/1399/

### CODE REVIEW

**Comment by kgpayne** [comment]
Testing in [tap-snowflake#15](https://github.com/MeltanoLabs/tap-snowflake/pull/15). Reduced total test run time from ~20m to ~2m 👏

**Comment by kgpayne** [comment]
More testing in [MeltanoLabs/tap-stackexchange#196](https://github.com/MeltanoLabs/tap-stackexchange/pull/196). Reduced test run time from ~8m to >1m ⏰

This is accomplished by adding an early exit to `get_records()` on the `RESTStream` subclass directly in the Tap, similar to the one in the SDK implementation of `SQLStream`. Applying this change to `RESTStream` in the SDK would resolve [#1349](https://github.com/meltano/sdk/issues/1349).

```python
# overridden to test _MAX_RECORDS_LIMIT
def get_records(self, context: dict | None) -> Iterable[dict[str, Any]]:
    """Return a generator of record-type dictionary objects.

    Each record emitted should be a dictionary of property names to their values.

    Args:
        context: Stream partition or context dictionary.

    Yields:
        One item per (possibly processed) record in the API.
    """
    index = 1
    for record in self.request_records(context):
        transformed_record = self.post_process(record, context)
        if transformed_record is None:
            # Record filtered out during post_process()
            continue

        if self._MAX_RECORDS_LIMIT and (index + 1) == self._MAX_RECORDS_LIMIT:
            # if we have reached the max records limit, return early
            break

        yield transformed_record
        index += 1
```

File: pyproject.toml [hunk 1]
```diff
@@ -175,6 +175,7 @@ omit = [
     "tests/*",
     "samples/*",
     "singer_sdk/helpers/_compat.py",
+    "singer_sdk/testing/*",
```
Comment (edgarrmondragon) [inline pyproject.toml:178]: Why ignore these for coverage?

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
I may be missing something but I don't see `get_records` using the limit or handling the exception.

**Comment by kgpayne** [comment]
@edgarrmondragon 

> I may be missing something but I don't see `get_records` using the limit or handling the exception.

That change isn't in this PR 😅 All I am doing here is setting `_MAX_RECORDS_LIMIT` on the streams after the Tap is instantiated and before testing begins. `SQLStream.get_records()` already reads `_MAX_RECORDS_LIMIT` and respects the limit, and there is an open PR in `tap-stackexchange` to test the same approach in `RESTStream.get_records()` (snippet [here](https://github.com/MeltanoLabs/tap-stackexchange/pull/196/files#diff-6ea50d44bd3c3d25fe6558f96d694fbdd50be1ac22d2a4ca6ac3c85a2e0e41a2R158-R183)).

I didn't want to conflate too many similar but separate changes in one PR, and `SQLStream`s already [stand to gain](https://github.com/MeltanoLabs/tap-snowflake/pull/15#issuecomment-1420962939) from this even if we never make the `RESTStream` change 🙂

File: pyproject.toml (hunk 1 above)
Comment (kgpayne) [reply pyproject.toml:178]: I didn't think it made sense to cover them as I can't see a way to easily test the test framework (as it relies on the pytest plugin) 🤔 I suppose we can test the individual tests, so I can roll back the ignore and create an issue for "testing the test cases". WDYT?

Comment (edgarrmondragon) [reply pyproject.toml:178]: Yeah, I think that's what we want eventually to be safe

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon
> 
> > I may be missing something but I don't see `get_records` using the limit or handling the exception.
> 
> That change isn't in this PR 😅 All I am doing here is setting `_MAX_RECORDS_LIMIT` on the streams after the Tap is instantiated and before testing begins. `SQLStream.get_records()` already reads `_MAX_RECORDS_LIMIT` and respects the limit, and there is an open PR in `tap-stackexchange` to test the same approach in `RESTStream.get_records()` (snippet [here](https://github.com/MeltanoLabs/tap-stackexchange/pull/196/files#diff-6ea50d44bd3c3d25fe6558f96d694fbdd50be1ac22d2a4ca6ac3c85a2e0e41a2R158-R183)).
> 
> I didn't want to conflate too many similar but separate changes in one PR, and `SQLStream`s already [stand to gain](https://github.com/MeltanoLabs/tap-snowflake/pull/15#issuecomment-1420962939) from this even if we never make the `RESTStream` change 🙂

Oh, ok. I've changed the title of the PR to better reflect that. Is there an issue to make `SuiteConfig.max_records_limit` apply universally?

**Review by aaronsteers — CHANGES_REQUESTED** [review]
Requesting hold to allow time to review.

**Comment by kgpayne** [comment]
@aaronsteers nudge 🙏 Would be good to get this out before it goes stale 🚢

File: pyproject.toml (hunk 1 above)
Comment (kgpayne) [reply pyproject.toml]: Done ✅

**Comment by aaronsteers** [comment]
I spent more time reviewing this, and I don't think we should merge in this current state.

Specifically, this doesn't seem to fit requirements of a [stable increment](https://handbook.meltano.com/engineering/stable-increments).

I'd suggest 2 changes:

1. The test function should be calling `run_connection_test()` which handles the error `MaxRecordsLimitException`.
2. We should consider renaming `_MAX_RECORDS_LIMIT` to `_DRY_RUN_RECORD_LIMIT` - or any other variable name which clearly informs developers (SDK contributors as well as tap developers) that this is _not_ a viable lever for forcing sync via small batches.

As I explain in #1349 and #1366, aborting the stream without reaching the end has to generate an error, and that error needs to be caught. Otherwise, state messages for the stream are not finalized - and cannot be finalized - 

It's important to note that developers may run these codepaths without a detailed understanding of implications - and we need to be careful that all options presented by the SDK still promise valid taps that comply with the Singer Spec. Exposing a records limit variable will _break_ compliance with the Singer Spec whenever that max limit is reached. 

There's a path forward here if we want to add this to a "partial success" return code, but the partial success return code needs to distinguish interruptions between these two categories: (1) streams that validly incremented state but had more records remaining, and (2) streams which did _not_ successfully reach a bookmark and therefor represent the start of an infinite loop, where no progress will ever be made towards syncing the stream successfully.

**Comment by aaronsteers** [comment]
I've addressed my own feedback above into this PR: #1436.

That PR combined with this one make the increment "stable", specifically in that we deal with the exceptions that are raised as a result of a max record count limit - including proper handling of streams according to the `replication_method` and whether or not the stream state is resumable.

Related, this proposes non-zero exit code options for taps "aborted" which were able to be successfully paused and left in a valid state despite leaving records in the source:

- #1409 

We don't need to tackle non-zero return codes as of now; this PR and #1436 both ensure that:

1. Test functions properly leverage max record abort thresholds.
2. Non-test functions will properly raise exceptions if record limit is misapplied by developers during "real" sync operations.

**Pull request closed without merging**